README.md

Gives polyline coordinates to set polylines in Google Map.

[![pub package](https://img.shields.io/pub/v/google_map_polyline.svg)](https://pub.dartlang.org/packages/google_map_polyline)

# google_map_polyline_new
Flutter plugin to retrieve coordinates (in latitude and longitude) to draw the polylines (Route) in `google_maps_flutter` package.
This package made with the inspiration of `flutter_polyline_points`.

This package is a continuation of the discontinued [google_map_polyline](https://pub.dev/packages/google_map_polyline)

### Getting Started

To use this plugin, add `google_map_polyline` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

### Example

1. First you need to create object using your **Google Maps API key**.
- Get an API key at [https://cloud.google.com/maps-platform/](https://cloud.google.com/maps-platform/).
	 (This plugin only requires **Directions API**.)

	```dart
		GoogleMapPolyline googleMapPolyline = 
			new  GoogleMapPolyline(apiKey:  "YOUR KEY HERE");
	```
		
2. Now you may able retrieve coordinates for the Polyline route.

- Get the Polyline coordinates using location coordinates (latitude and longitude)
	 ```dart
     	await googleMapPolyline.getCoordinatesWithLocation(
     		origin: LatLng(40.677939, -73.941755),
     		destination: LatLng(40.698432, -73.924038),
     		mode:  RouteMode.driving);	
	```


- Get the Polyline coordinates using address of the location
	```dart
		await googleMapPolyline.getPolylineCoordinatesWithAddress(
			origin:  '55 Kingston Ave, Brooklyn, NY 11213, USA',
			destination:  '8007 Cypress Ave, Glendale, NY 11385, USA',
			mode:  RouteMode.driving);
	```

### Overview
- [x] Coordinates with Location Coordinates (Latitude and Longitude)
- [x] Coordinates with Location Address
- [x] Route Modes
  	- [x] Driving
  	- [x] Walking
  	- [x] Bicycling
- [ ] Coordinates with Location Coordinates and Address (example: Origin as address and Destination as coordinates)
- [ ] Alternative routes


### Feature Requests and Issues
Please file feature requests and bugs at the  [issue tracker](https://github.com/Shark01/google_map_polyline/issues/new).


CHANGELOG.md
## [0.0.1 - 0.0.3+1]
- Initial publish

## [0.0.4]
- Updated dependencies

## [0.2.0 - 0.2.0+1]
- Minor changes

## [0.3.0+1]
- Fixing Deprecated Android Embedding v2
- Migrating to Null Safety

## [0.3.0+2]
- Updating file name path and some fix in exception

lib/src/map_polyline.dart
import 'package:google_map_polyline_new/src/polyline_request.dart';
import 'package:google_map_polyline_new/src/polyline_utils.dart';
import 'package:google_map_polyline_new/src/route_mode.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
/// Gives polyline coordinates to set polylines in Google Map.
class GoogleMapPolyline {
  String apiKey;
  GoogleMapPolyline({required this.apiKey});

  late PolylineUtils _utils;
  PolylineRequestData? _data;

  /// Get coordinates using Location Coordinates
  /// Example : LatLng(40.677939, -73.941755)
  Future<List<LatLng>?> getCoordinatesWithLocation({
    required LatLng origin,
    required LatLng destination,
    required RouteMode mode,
  }) async {
    _data = new PolylineRequestData(
        originLoc: origin,
        destinationLoc: destination,
        mode: mode,
        locationText: false,
        apiKey: apiKey);

    _utils = new PolylineUtils(_data);

    return await _utils.getCoordinates();
  }

  /// Get coordinates using Location Coordinates
  /// Example : '55 Kingston Ave, Brooklyn, NY 11213, USA'
  Future<List<LatLng>?> getPolylineCoordinatesWithAddress({
    required String origin,
    required String destination,
    required RouteMode mode,
  }) async {
    _data = new PolylineRequestData(
        originText: origin,
        destinationText: destination,
        mode: mode,
        locationText: true,
        apiKey: apiKey);

    _utils = new PolylineUtils(_data);

    return await _utils.getCoordinates();
  }
}


lib/src/polyline_utils.dart
import 'package:dio/dio.dart';
import 'package:google_map_polyline_new/src/route_mode.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:google_map_polyline_new/src/polyline_request.dart';

class PolylineUtils {
  PolylineRequestData? _data;
  PolylineUtils(this._data);

  Future<List<LatLng>?> getCoordinates() async {
    List<LatLng>? _coordinates = [];

    var qParam = {
      'mode': getMode(_data!.mode),
      'key': _data!.apiKey,
    };

    if (_data!.locationText!) {
      qParam['origin'] = _data!.originText;
      qParam['destination'] = _data!.destinationText;
    } else {
      qParam['origin'] =
          "${_data!.originLoc!.latitude},${_data!.originLoc!.longitude}";
      qParam['destination'] =
          "${_data!.destinationLoc!.latitude},${_data!.destinationLoc!.longitude}";
    }

    Response _response;
    Dio _dio = new Dio();
    _response = await _dio.get(
        "https://maps.googleapis.com/maps/api/directions/json",
        queryParameters: qParam);

    try {
      if (_response.statusCode == 200) {
        if (_response.data["status"] == "REQUEST_DENIED") {
          throw Exception(_response.data["error_message"]);
        }

        _coordinates = decodeEncodedPolyline(
            _response.data['routes'][0]['overview_polyline']['points']);
      }
    } catch (e) {
      print('[ERROR_POLYLINE] ${e.toString()}');
    }

    return _coordinates;
  }

  List<LatLng> decodeEncodedPolyline(String encoded) {
    List<LatLng> poly = [];
    int index = 0, len = encoded.length;
    int lat = 0, lng = 0;

    while (index < len) {
      int b, shift = 0, result = 0;
      do {
        b = encoded.codeUnitAt(index++) - 63;
        result |= (b & 0x1f) << shift;
        shift += 5;
      } while (b >= 0x20);
      int dlat = ((result & 1) != 0 ? ~(result >> 1) : (result >> 1));
      lat += dlat;

      shift = 0;
      result = 0;
      do {
        b = encoded.codeUnitAt(index++) - 63;
        result |= (b & 0x1f) << shift;
        shift += 5;
      } while (b >= 0x20);
      int dlng = ((result & 1) != 0 ? ~(result >> 1) : (result >> 1));
      lng += dlng;
      LatLng p = new LatLng((lat / 1E5).toDouble(), (lng / 1E5).toDouble());
      poly.add(p);
    }
    return poly;
  }

  String? getMode(RouteMode? _mode) {
    switch (_mode) {
      case RouteMode.driving:
        return 'driving';
      case RouteMode.walking:
        return 'walking';
      case RouteMode.bicycling:
        return 'bicycling';
        case RouteMode.transit:
        return 'transit';
      default:
        return null;
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:google_map_polyline_new/google_map_polyline_new.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int _polylineCount = 1;
  Map<PolylineId, Polyline> _polylines = <PolylineId, Polyline>{};

  GoogleMapPolyline _googleMapPolyline =
      new GoogleMapPolyline(apiKey: "YOUR KEY HERE");

  //Polyline patterns
  List<List<PatternItem>> patterns = <List<PatternItem>>[
    <PatternItem>[], //line
    <PatternItem>[PatternItem.dash(30.0), PatternItem.gap(20.0)], //dash
    <PatternItem>[PatternItem.dot, PatternItem.gap(10.0)], //dot
    <PatternItem>[
      //dash-dot
      PatternItem.dash(30.0),
      PatternItem.gap(20.0),
      PatternItem.dot,
      PatternItem.gap(20.0)
    ],
  ];

  LatLng _mapInitLocation = LatLng(40.683337, -73.940432);

  LatLng _originLocation = LatLng(40.677939, -73.941755);
  LatLng _destinationLocation = LatLng(40.698432, -73.924038);

  bool _loading = false;

  _onMapCreated(GoogleMapController controller) {
    setState(() {
    });
  }

  //Get polyline with Location (latitude and longitude)
  _getPolylinesWithLocation() async {
    _setLoadingMenu(true);
    List<LatLng>? _coordinates =
        await _googleMapPolyline.getCoordinatesWithLocation(
            origin: _originLocation,
            destination: _destinationLocation,
            mode: RouteMode.driving);

    setState(() {
      _polylines.clear();
    });
    _addPolyline(_coordinates);
    _setLoadingMenu(false);
  }

  //Get polyline with Address
  _getPolylinesWithAddress() async {
    _setLoadingMenu(true);
    List<LatLng>? _coordinates =
        await _googleMapPolyline.getPolylineCoordinatesWithAddress(
            origin: '55 Kingston Ave, Brooklyn, NY 11213, USA',
            destination: '8007 Cypress Ave, Glendale, NY 11385, USA',
            mode: RouteMode.driving);

    setState(() {
      _polylines.clear();
    });
    _addPolyline(_coordinates);
    _setLoadingMenu(false);
  }

  _addPolyline(List<LatLng>? _coordinates) {
    PolylineId id = PolylineId("poly$_polylineCount");
    Polyline polyline = Polyline(
        polylineId: id,
        patterns: patterns[0],
        color: Colors.blueAccent,
        points: _coordinates!,
        width: 10,
        onTap: () {});

    setState(() {
      _polylines[id] = polyline;
      _polylineCount++;
    });
  }

  _setLoadingMenu(bool _status) {
    setState(() {
      _loading = _status;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      darkTheme: ThemeData(brightness: Brightness.dark),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Map Polyline'),
        ),
        body: Container(
          child: LayoutBuilder(
            builder: (context, cont) {
              return Column(
                children: <Widget>[
                  SizedBox(
                    width: MediaQuery.of(context).size.width,
                    height: MediaQuery.of(context).size.height - 175,
                    child: GoogleMap(
                      onMapCreated: _onMapCreated,
                      polylines: Set<Polyline>.of(_polylines.values),
                      initialCameraPosition: CameraPosition(
                        target: _mapInitLocation,
                        zoom: 15,
                      ),
                    ),
                  ),
                  Expanded(
                    child: Align(
                      alignment: Alignment.center,
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                        children: <Widget>[
                          ElevatedButton(
                              onPressed: _getPolylinesWithLocation,
                              child: Text('Polylines with Location')),
                          ElevatedButton(
                              onPressed: _getPolylinesWithAddress,
                              child: Text('Polylines with Address'))
                        ],
                      ),
                    ),
                  ),
                ],
              );
            },
          ),
        ),
        floatingActionButtonLocation: FloatingActionButtonLocation.centerDocked,
        floatingActionButton: _loading
            ? Container(
                color: Colors.black.withOpacity(0.75),
                child: Center(
                  child: Text(
                    'Loading...',
                    style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
                  ),
                ),
              )
            : Container(),
      ),
      debugShowCheckedModeBanner: false,
    );
  }
}
